### DISCUSSION TITLE
Money amount: precision based on currency

### DISCUSSION THREAD

**Discussion by mdomke** [sqlalchemy/sqlalchemy#7124] [Usage Questions] [answered]
I have a model with two fields `amount` and `currency`. The `amount` field is currently a custom type, which handles `Decimal` values but persists them as integers. The `Decimal` type looks something like this:
```python
class Decimal(sa.types.TypeDecorator):
    impl = sa.types.INTEGER

    # This was computed by `math.log(2 ** 64, 10) - 1`. This is the maximum power of 10 we can
    # store in a 64 bit integer.
    MAX_POW = 18

    def __init__(self, digits, decimal_places):
        if digits < 0 or decimal_places < 0:
            raise ValueError("Can't have negative amounts of digits or decimal places")
        if (digits + decimal_places) > self.MAX_POW:
            raise ValueError("We can't store such precision")
        super().__init__()
        self.digits = digits
        self.decimal_places = decimal_places

    def process_bind_param(self, value, dialect):
        if value is None:
            return value

        if isinstance(value, int):
            value = decimal.Decimal(value)
        if not isinstance(value, decimal.Decimal):
            raise TypeError("Values are expected to be python Decimals")

        if count_integer_part(value) > self.digits:
            raise ValueError("Can't store that many digits before the decimal point")
        if count_fractional_part(value) > self.decimal_places:
            raise ValueError("Can't store that many digits after the decimal point")

        integer = value.scaleb(self.decimal_places)
        assert integer % 1 == 0, "Integer should not have digits after the decimal point"
        return int(integer)

    def process_result_value(self, value, dialect):
        if value is None:
            return value
        return decimal.Decimal(value).scaleb(-self.decimal_places)

    def __repr__(self):
        return f"Decimal({self.digits}, {self.decimal_places})"
```
Currently all fields holding money values use the parametrised `MoneyAmount`
```
MoneyAmount = functools.partial(Decimal, digits=16, decimal_places=2)
```
I now want to support currencies with **more than two** decimal places. The actual number of decimal places should be calculated from the `currency` field with an appropriate lookup-table. I have been trying to use a [hybrid_property](https://docs.sqlalchemy.org/en/14/orm/extensions/hybrid.html#module-sqlalchemy.ext.hybrid), but have some trouble defining the `hybrid_property.expression` (or `hybrid_property.comparator`), I would have to define the same lookup in a SQL expression.
```python
decimal_places: dict[str, int] = {
    "BHD": 3,
}

class Table(declarative_base()):
    __tablename__ = "test"

    def __init__(self, **kwargs):
        amount = kwargs.pop("amount", None)
        super().__init__(**kwargs)
        if amount is not None:
            self.amount = amount

    id = sa.Column(sa.Integer, primary_key=True)
    _amount = sa.Column(sa.Integer, nullable=False, default=0)
    currency = sa.Column(sa.String(3), nullable=False, default="")

    @hybrid_property
    def amount(self):
        places = decimal_places.get(self.currency, 2)
        return Decimal(self._amount).scaleb(-places)

    @amount.setter
    def amount(self, value):
        if not isinstance(value, Decimal):
            value = Decimal(value)
        places = decimal_places.get(self.currency, 2)
        integer = value.scaleb(places)
        self._amount = int(integer)

    @amount.expression
    def amount(cls):
        # TODO: What to do here?
```
I found [this](https://techspot.zzzeek.org/2011/10/29/value-agnostic-types-part-ii/) old blog post that is somehow dealing with the same problem domain, but uses numeric/float types for storing data. 
Do you guys have any suggestion on how to approach this problem?

**Edit**: I actually cross-posted this question on [Stackoverflow](https://stackoverflow.com/questions/69397131/sqlalchemy-money-amount-precision-based-on-currency). I hope this is not considered bad style!

**Comment by zzzeek** [comment]
so...the hybrid version is meant to replace the TypeDecorator version?   for the expression part, if you use `@expression` I think you still need to use the TypeDecorator you have, and return _amount as your Decimal type. Because the thing you want to have here is "Table.amount == Decimal('1234.56')", and the right hand side has to be correctly coerced to an integer for the DB part of it.  You either need the custom type there, or you need to forego using `@expression` and use the `@comparator` instead: https://docs.sqlalchemy.org/en/14/orm/extensions/hybrid.html#building-custom-comparators .

I think if you keep _amount as your Decimal type, your hybrid only need coerce the incoming value to your Decimal type, and for the SQL part can send out the column as it is, the comparison logic would be handled by process_bind_param().

**Reply by mdomke** [reply]
I don't necessarily want to replace the `TypeDecorator` version, but as far as I can see there is no way for the `Decimal`-type to dynamically look up the number of decimal places, by peeking into the value of the `currency`-field. That's why I was trying to understand if the `hybrid_property` could be helpful in that regard.
If I left the `_amount`-field to be of type `MoneyAmount` with a fixed amount of decimal places (in this case 2), values with a higher precision could not be persisted. So, the main problem is that the `Decimal`/`MoneyAmount` field is not able to adjust the interpretation of the underlying integer value on the fly.
I think I will work around the problem, by adjusting the number of decimal places for the `MoneyAmount` to, let's say, 4 in general so that it can accommodate money amounts in all relevant currencies. I will then take in the view part of the application to crop the number of decimal places accordingly. Thank you very much for your help, Mike!

**Comment by gordthompson** [comment, marked as answer]
> I think I will work around the problem, by adjusting the number of decimal places for the MoneyAmount to, let's say, 4 in general so that it can accommodate money amounts in all relevant currencies. I will then take in the view part of the application to crop the number of decimal places accordingly.

That's what I would suggest, except that I would add a couple of extra decimal places. Many applications that cater to US-style currency (2 decimal places) use DECIMAL(18,4) as their data type to avoid rounding errors in calculations. Discussions like this

https://www.quora.com/Does-any-countrys-currency-have-more-than-two-decimal-digits

suggest that three decimal places is the (current) maximum, so personally I would go with 4 + 2 "extra" = 6 decimal places in total.

And you are right in recognizing that the *storage issue* and the *display issue* are separate concerns.
